File: 438-find-all-anagrams-in-a-string/438-find-all-anagrams-in-a-string.py
class Solution:
    def findAnagrams(self, s: str, p: str) -> List[int]:
        
        
        sl = len(s)
        pl = len(p)
        
        if pl > sl:
            return 
        
        pCount = [0]*26
        sCount = [0]*26
        
        for e in p:
            pCount[ord(e)-ord('a')] +=1 
        
        ret = [] 
        for i in range(sl):
            
            
            sCount[ord(s[i]) -ord('a')] += 1
            
            if i >=pl:
                sCount[ord(s[i-pl]) -ord('a')]  -=1
                
            if sCount ==pCount:
                ret.append( i-pl+1)
        return ret 
        

File: 438-find-all-anagrams-in-a-string/438-find-all-anagrams-in-a-string.cpp
class Solution {
public:
    vector<int> findAnagrams(string s, string p) {
        
        

      int sl = s.length();
      int pl = p.length();

      if (s.size() < p.size()) return {};

      vector<int> palpha(26,0);
      vector<int> window(26,0);

      for(int i=0; i <pl ; i++){
        palpha[p[i] -'a'] ++;
        window[s[i]- 'a'] ++;
      }
      vector<int> ret;
      if(palpha == window) ret.push_back(0);

      for(int i=pl ; i <sl ; i++){
        window[s[i-pl] - 'a'] --;
        window[s[i] -'a'] ++;

        if ( palpha == window){
          ret.push_back(i-pl+1);
        }
      }
        
    return ret; 

    }
};